oidc_provider/models.py

```diff
@@ -161,6 +167,49 @@ def scope(self, value):
     def default_redirect_uri(self):
         return self.redirect_uris[0] if self.redirect_uris else ''
 
+    def is_allowed_redirect_uri(self, redirect_uri_str):
+        """
+        Determine whether a given `redirect_uri` is allowed for this client.
+
+        This uses simple string matching UNLESS the host section is equal to
+        either '127.0.0.1' or '[::1]'. This is to support native apps which
+        bind to an ephemeral port.
+
+        See https://tools.ietf.org/html/rfc8252#section-7.3
+        """
+
+        if redirect_uri_str in self.redirect_uris:
+            return True
+
+        try:
+            redirect_uri = urlparse(
+                redirect_uri_str if "//" in redirect_uri_str else "//" + redirect_uri_str
+            )
+        except ValueError:
+            return False
+
+        if redirect_uri.hostname not in ["::1", "127.0.0.1"]:
+            return False
+
+        for allowed_redirect_uri_str in self.redirect_uris:
+            try:
+                allowed_redirect_uri = urlparse(
+                    allowed_redirect_uri_str if "//" in allowed_redirect_uri_str
+                    else "//" + allowed_redirect_uri_str
+                )
+            except ValueError:
+                return False
```

Comment:
The way this check works, if there are any URI's that cannot be parsed, in the allowed list, before the URI that matches the loopback URI it will return False. That feels unintuitive to me.

I couldn't find any examples of URIs that raise ValueErrors though. In practice, I don't think this is a real problem.

Reply:
[`urlparse`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) defines the cases that raise `ValueError`.

I'll change it to `continue` if a `ValueError` is raised.

Reply:
It seemed like a narrow set of circumstances, I wasn't able to create anything that actually decomposed to cause the ValueError. The port issue only comes up when evaluating the port (which is ignored on purpose in this code).

It was only the square bracket that I managed to get it to happen with.

Changing to continue sounds good though, I think that makes sense.